Improve auto-atomicity for conditionals

stephentoub opened the pull request:
- Allow for automatically making loops that end conditional branches be atomic
- Allow for using a following expression conditional to make a prior loop atomic
- Allow conditionals to influence the computed minimum length

Fixes https://github.com/dotnet/runtime/issues/62723
Contributes to https://github.com/dotnet/runtime/issues/62449

danmoseley commented on src/libraries/System.Text.RegularExpressions/tests/RegexReductionTests.cs:
```diff
@@ -518,6 +542,16 @@ public void PatternsReduceDifferently(string pattern1, string pattern2)
+        [InlineData(@"(abc)(?(1)\d{3}|\d{2})", 5)]
+        [InlineData(@"(abc)(?(1)\d{2}|\d{3})", 5)]
+        [InlineData(@"(abc)(?(1)|\d{2})", 3)]
+        [InlineData(@"(abc)(?(1)\d{3})", 3)]
```
do we have 
```c#
[InlineData(@"(abc|)", 0)]
[InlineData(@"(|abc)", 0)]
[InlineData(@"(?(x)abc|)", 0)]
[InlineData(@"(?(x)|abc)", 0)]
```
or similar

stephentoub replied:
Those are just two branches, one of which is Empty.  I can add them if you think it's valuable, but I also would be very surprised if those failed and all of the other tests passed.

danmoseley replied:
channeling my edge case tester DNA. sure, as you wish.

stephentoub replied:
Already added :)

danmoseley commented on src/libraries/System.Text.RegularExpressions/tests/RegexReductionTests.cs:
```diff
@@ -371,6 +371,17 @@ private static int GetMinRequiredLength(Regex r)
         [InlineData("xyz(?:(?i:abcde)|(?i:abcdf))", "xyz(?i:abcd[ef])")]
         [InlineData("bonjour|hej|ciao|shalom|zdravo|pozdrav|hallo|hola|hello|hey|witam|tere|bonjou|salam|helo|sawubona", "(?>bonjou(?>r|)|h(?>e(?>j|(?>l(?>lo|o)|y))|allo|ola)|ciao|s(?>halom|a(?>lam|wubona))|zdravo|pozdrav|witam|tere)")]
         [InlineData("\\w\\d123|\\w\\dabc", "\\w\\d(?:123|abc)")]
+        // Loops inside alternation constructs
```
are there enough tests ensuring this doesn't happen if there's incompatible options? eg this should not atomify:
`(a(?-i:b)(?-i:B*)|def)ghi`
and possibly the same for other options. I don't see `(?-i` in this file at all.

danmoseley replied:
With respect to options, perhaps these should have a optional global option. So
`[InlineData("abB*", "ab+", RegexOptions.IgnoreCase]` should pass?

stephentoub replied:
> With respect to options, perhaps these should have a optional global option

The options check / impact will go away https://github.com/dotnet/runtime/issues/61048; since we're ignoring RightToLeft, the only other option that's relevant here (and the reason this check exists) is IgnoreCase, and after #61048, IgnoreCase will be removed from everything except for backreferences.

danmoseley replied:
Ah, yes.  Unrelated, it does seem we're a little short on functional tests of inline enable/disable of options, I'll make a note.

stephentoub committed:
Improve auto-atomicity for conditionals

- Allow for automatically making loops that end conditional branches be atomic
- Allow for using a following expression conditional to make a prior loop atomic
- Allow conditionals to influence the computed minimum length
